infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -374,28 +424,40 @@ public boolean isServerReadyToHandleMsg(CorfuMsg msg) {
     private synchronized void resetLogUnit(CorfuPayloadMsg<Long> msg,
                                            ChannelHandlerContext ctx, IServerRouter r) {
 
+        if (msg.getPayload() <= serverContext.getLogUnitEpochWaterMark()
+                || msg.getPayload() != serverContext.getServerEpoch()) {
+            log.info("LogUnit Server Reset request received but reset already done.");
+            r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+            return;
+        }
+
         // Check if the reset request is with an epoch greater than the last reset epoch seen to
-        // prevent multiple reset in the same epoch. and should be equal to the current router
+        // prevent multiple reset in the same epoch and should be equal to the current router
         // epoch to prevent stale reset requests from wiping out the data.
-        if (msg.getPayload() > serverContext.getLogUnitEpochWaterMark()
-                && msg.getPayload() == serverContext.getServerEpoch()) {
-            serverContext.setLogUnitEpochWaterMark(msg.getPayload());
-            batchWriter.addTask(RESET, msg)
+        serverContext.setLogUnitEpochWaterMark(msg.getPayload());
+
+        Lock lock = resetLock.writeLock();
+        lock.lock();
```

Comment:
Locking here is bad, the reset op is added to BatchProcessor, but will be run after all previous write operations in the queue are finished, during that all reads will be blocked. This is a performance issue.

Also if you put lock into streamLogFiles then you don't need to do cf.join() below.

Reply:
done